Specify components in `rust-toolchain` and update the toolchain to `nightly-2020-08-20`.

jrvanwhy opened the pull request:
I had to update `linked_list_allocator` because it is not compatible with the newer compiler (`const_fn` changes). I also silenced a warning on the `start` definition.

jrvanwhy commented:
Converting to draft for two reasons:

1. `master`'s build is broken because a new version of the `log` crate was pushed to crates.io so now the toolchain version increase is a bugfix. I'm pulling that change out into a new PR and keeping the `rust-toolchain` components change in this PR.
2. Our CI uses an outdated version of `rustup` that doesn't support the new format -- I'm going to look into how easy or difficult this is to fix.

hudson-ayers commented:
Any reason for only advancing to 08-20 instead of matching the kernel at 2020-10-25?

Not a big deal, just curious if newer breaks anything

jrvanwhy commented:
> Any reason for only advancing to 08-20 instead of matching the kernel at 2020-10-25?
> 
> Not a big deal, just curious if newer breaks anything

Yeah, newer breaks stuff:

```rust
error: could not compile `libtock_core`

Caused by:
  process didn't exit successfully: `rustc --crate-name empty_main --edition=2018 core/examples/empty_main.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C opt-level=3 -C panic=abort -C lto -C debuginfo=2 -C metadata=8c9e05283eac2f23 -C extra-filename=-8c9e05283eac2f23 --out-dir /home/jrvanwhy/libtock-rs/target/thumbv7em-none-eabi/release/examples --target thumbv7em-none-eabi -L dependency=/home/jrvanwhy/libtock-rs/target/thumbv7em-none-eabi/release/deps -L dependency=/home/jrvanwhy/libtock-rs/target/release/deps --extern libtock_codegen=/home/jrvanwhy/libtock-rs/target/release/deps/liblibtock_codegen-8d52739790490cf4.so --extern libtock_core=/home/jrvanwhy/libtock-rs/target/thumbv7em-none-eabi/release/deps/liblibtock_core-4f170171c9f158bc.rlib --extern libtock_platform=/home/jrvanwhy/libtock-rs/target/thumbv7em-none-eabi/release/deps/liblibtock_platform-2d42bde2cd607522.rlib -D warnings -C relocation-model=static -C link-arg=-Tlayout.ld` (signal: 11, SIGSEGV: invalid memory reference)
```

I didn't try to narrow down the exact release where the breakage appeared, as it was taking time (we also need both `miri` and `rustfmt`, which aren't in every nightly). As I am in the middle of rewriting a lot of `libtock_core`, I didn't feel the need to look hard into the failure.

jrvanwhy commented:
The (next? current?) version of GitHub Actions' Ubuntu 18.04 environment will have `rustup` 1.23.0, which supports the `rust-toolchain` file in this PR ([Source](https://github.com/actions/virtual-environments/pull/2159)). It looks like Actions environment updates roll out frequently, so I expect we'll see the new image appear within the next couple of weeks. As a result, my plan for fixing the `rustup` version mismatch is to wait for that update.

The other issue I named -- the broken build in master -- was fixed in #264

jrvanwhy commented:
GitHub has now rolled out the Ubuntu 18.04 environment with `rustup` 1.23.0, but `actions-rs/toolchain` needs work in order to support the new `rust-toolchain` format. However, `rustup` has now implemented all features that we were using from `actions-rs/toolchain`, so I removed our use of `actions-rs/toolchain` and CI passes again!

Our `size-diff` workflow failed because it tries to build `master` and I removed the `rustup target add` invocation that `master` needs to build. This is a one-off failure that won't happen again when this PR is merged.

jrvanwhy commented:
I believe this PR is ready to merge. You'll need to merge manually because size-diff is broken (as explained before, it'll be fixed once this is in master).
